File: 0001-two-sum/0001-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # TC: O(N)
        # SC: O(N)
        
        num_dict = {}

        for i, num in enumerate(nums):
            complement = target - num

            if complement in num_dict:
                return [i, num_dict[complement]]
            
            num_dict[num] = i
        
        return []

File: 0001-two-sum/0001-two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // TC: O(N)
        // SC: O(N)
        
        unordered_map<int, int> hashset;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            int complement = target - nums[i];
            if (hashset.contains(complement)) {
                return {hashset[complement], i};
            }
            hashset[nums[i]] = i;
        }

        return {};
    }
};